uboot-envtools: set nonshared flag correctly

map-b opened the pull request:
Currently, uboot-envtools is being built for a (shared) instruction set (phase2 in buildbots) instead of target-specific (phase1 in builbots). So the package does not contain the uci-defaults file specific for each target_subtarget. (Only the fortunate target with coincidentally the same instruction set as the SDK chosen by the buildbot will have the uci-defaults file.)

This commit sets the nonshared flag correctly in the Makefile so that uboot-envtools is built for the target-specific.

Fixes: #19040
Fixes: 46e376c ("uboot-tools: migrate uboot-envtools to uboot-tools")

hauke reviewed (approved):
LGTM. This was introduced in main branch and does not affect 24.10

This is the only package missing between these two:
https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/packages/
https://downloads.openwrt.org/releases/24.10.1/targets/mediatek/filogic/packages/

map-b commented:
The first attempt does not work neither, I need to define `PKG_FLAGS` before the `package.mk` include. https://github.com/openwrt/openwrt/blob/a51359aafde18b7e809842b4b4c31c78f5fa0c51/package/boot/uboot-tools/Makefile#L23 which includes `package-defaults.mk` https://github.com/openwrt/openwrt/blob/a51359aafde18b7e809842b4b4c31c78f5fa0c51/include/package.mk#L137 where the final `PKGFLAGS` is set. https://github.com/openwrt/openwrt/blob/a51359aafde18b7e809842b4b4c31c78f5fa0c51/include/package-defaults.mk#L34-L35
Then, in `package-dumpinfo.mk`, the flag `nonshared` is checked and the `Repository` field is set  https://github.com/openwrt/openwrt/blob/a51359aafde18b7e809842b4b4c31c78f5fa0c51/include/package-dumpinfo.mk#L32 (here, the file `tmp/.packageinfo` is generated after running `make menuconfig/defconfig`)

Finally in `scripts/package-metadata.pl` that `Repository` field is checked to add the package to the `ALL_NONSHARED` set.
https://github.com/openwrt/openwrt/blob/a51359aafde18b7e809842b4b4c31c78f5fa0c51/scripts/package-metadata.pl#L310-L319

 ------------------------------------
So, **moving the defining there implies that the all three packages in the uboot-tools package will be affected with the `nonshared` flag** (Each of the two extra packages are 70KB aprox.)

I tried modifying `PKGFLAGS` directly (instead of `PKG_FLAGS`) and it seems to work per sub-package  but I am not sure if there could be unexpected consequences.
```
diff --git a/package/boot/uboot-tools/Makefile b/package/boot/uboot-tools/Makefile
index aaf5c0afee..9a630c8961 100644
--- a/package/boot/uboot-tools/Makefile
+++ b/package/boot/uboot-tools/Makefile
@@ -19,7 +19,6 @@ PKG_LICENSE:=GPL-2.0 GPL-2.0+
 PKG_LICENSE_FILES:=Licenses/README
 
 PKG_BUILD_PARALLEL:=1
-PKG_FLAGS:=nonshared
 
 include $(INCLUDE_DIR)/package.mk
 
@@ -54,6 +53,7 @@ define Package/uboot-envtools
 	SECTION:=utils
 	CATEGORY:=Utilities
 	SUBMENU:=Boot Loaders
+	PKGFLAGS+=nonshared
 	TITLE:=read/modify U-Boot bootloader environment
 	URL:=http://www.denx.de/wiki/U-Boot
 endef
```
(*) My tests are done checking the file `tmp/.packageinfo` looking at the field `Repository`. Only removing that file and calling `make defconfig-menuconfig` is enough to regenerate the file and retest.

map-b commented:
> LGTM. This was introduced in main branch and does not affect 24.10
> 
> This is the only package missing between these two: https://downloads.openwrt.org/snapshots/targets/mediatek/filogic/packages/ https://downloads.openwrt.org/releases/24.10.1/targets/mediatek/filogic/packages/

This flag also is not taken into account?? but it a special case so I do not know. https://github.com/openwrt/openwrt/blob/a51359aafde18b7e809842b4b4c31c78f5fa0c51/package/kernel/linux/Makefile#L57

map-b commented:
So it seems that using PKGFLAGS is the right way to proceed. This pattern of PKG_<FIELD>, <FIELD> like PKG_FLAGS, PKGFLAGS, PKG_URL, URL, PKG_MAINTAINER, MAINTAINER, … was created to be able to define a global scope(PKG_<FIELD>) or a binary scope ( <FIELD>) fields. 
https://github.com/openwrt/openwrt/blob/583f81ef0ee152404b633810607bbe15b0dcf9cf/include/package-defaults.mk#L13-L60
For reference: 349e7b635 , 2d7eaf2e15 , 064e7c8f00, 2cb75cd8b9 and e32edf712b

map-b commented:
Also for reference for future visitors, this is the logic behind which packages are built in **phase 1** or **phase 2**, basically it depends in which directories are put depending of the `Repository:` field. https://github.com/openwrt/openwrt/commit/5170393f8cae3c5acac61bd6ac6e127f9727cee4

map-b committed:
uboot-envtools: set nonshared flag correctly per-subpackage

Currently, uboot-envtools is being built for a (shared) instruction
set (phase2 in buildbots) instead of target-specific (phase1 in builbots).
So the package does not contain the uci-defaults file specific for each
target_subtarget. (Only the fortunate target with coincidentally the same
instruction set and target as the SDK chosen by the buildbot will have the
uci-defaults file.)

This commit sets the nonshared flag correctly in the Makefile so that
uboot-envtools is built for the target-specific (phase 1).

This is done by using the PKGFLAGS variable which is intended for
per-binary flags (instead of PKG_FLAGS which has a global scope), as
stated in some old commits 349e7b635e[1], 2d7eaf2e15[2], 064e7c8f00[3] and
2cb75cd8b9[4].

While at here, use PKG_URL, instead of URL, which is intended for global
scope. As stated in commit e32edf712b[5].

[1] https://github.com/openwrt/openwrt/commit/349e7b635ea819ad34a571bc0adcce9b205df4e9
[2] https://github.com/openwrt/openwrt/commit/2d7eaf2e15aa77bcf526b7ea9747732b35194b2e
[3] https://github.com/openwrt/openwrt/commit/064e7c8f00ebc1721553c0fcb312f809b00e179a
[4] https://github.com/openwrt/openwrt/commit/2cb75cd8b9d451bde0cfffce80fa67d58944c44b
[5] https://github.com/openwrt/openwrt/commit/e32edf712b7c43e25b4bec40fedfd948a0ed5bd6

Fixes: #19040
Fixes: 46e376c ("uboot-tools: migrate uboot-envtools to uboot-tools")

Signed-off-by: Mario Andrés Pérez <mapb_@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/19180
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Thanks! Rebased on top of main and merged!
